File: packages/schemas/python/src/app_schemas/resume.py
"""Résumé / profile contract — the product's single source of truth.

JSON Resume-aligned (https://jsonresume.org/schema) so the same data also works
with that ecosystem's themes and tooling. Pydantic v2; mirrors the zod shapes in
``packages/schemas/src/resume.ts`` field-for-field.
"""

from __future__ import annotations

from datetime import date

from pydantic import BaseModel, ConfigDict, EmailStr, Field, HttpUrl


class _Model(BaseModel):
    # Accept both JSON Resume camelCase (aliases) and snake_case; ignore extras.
    model_config = ConfigDict(populate_by_name=True, extra="ignore")


class Location(_Model):
    city: str | None = None
    region: str | None = None
    country_code: str | None = Field(default=None, alias="countryCode")


class ProfileLink(_Model):
    network: str
    username: str | None = None
    url: HttpUrl


class Basics(_Model):
    name: str
    label: str | None = None  # headline / current title
    email: EmailStr | None = None
    phone: str | None = None
    url: HttpUrl | None = None
    summary: str | None = None
    location: Location | None = None
    profiles: list[ProfileLink] = Field(default_factory=list)


class Work(_Model):
    name: str  # company
    position: str
    url: HttpUrl | None = None
    location: str | None = None
    start_date: str | None = Field(default=None, alias="startDate")
    end_date: str | None = Field(default=None, alias="endDate")
    summary: str | None = None
    highlights: list[str] = Field(default_factory=list)
    tech: list[str] = Field(default_factory=list)


class Education(_Model):
    institution: str
    area: str | None = None
    study_type: str | None = Field(default=None, alias="studyType")
    start_date: str | None = Field(default=None, alias="startDate")
    end_date: str | None = Field(default=None, alias="endDate")
    score: str | None = None
    courses: list[str] = Field(default_factory=list)


class Skill(_Model):
    name: str
    level: str | None = None
    keywords: list[str] = Field(default_factory=list)


class Project(_Model):
    name: str
    description: str | None = None
    url: HttpUrl | None = None
    highlights: list[str] = Field(default_factory=list)
    keywords: list[str] = Field(default_factory=list)


class Certificate(_Model):
    name: str
    date: str | None = None
    issuer: str | None = None
    url: HttpUrl | None = None


class Meta(_Model):
    last_updated: date | None = Field(default=None, alias="lastUpdated")
    version: str | None = None


class Resume(_Model):
    """Full résumé document — the canonical shape served + rendered + exported."""

    basics: Basics
    work: list[Work] = Field(default_factory=list)
    education: list[Education] = Field(default_factory=list)
    skills: list[Skill] = Field(default_factory=list)
    projects: list[Project] = Field(default_factory=list)
    certificates: list[Certificate] = Field(default_factory=list)
    meta: Meta | None = None


# `Profile` is an alias for the public-facing résumé document.
Profile = Resume

File: packages/schemas/src/resume.ts
import { z } from "zod";

// Résumé / profile contract — the product's single source of truth.
// JSON Resume-aligned (https://jsonresume.org/schema). Mirrors the Pydantic
// models in packages/schemas/python/src/app_schemas/resume.py field-for-field.

export const Location = z.object({
  city: z.string().optional(),
  region: z.string().optional(),
  countryCode: z.string().optional(),
});
export type Location = z.infer<typeof Location>;

export const ProfileLink = z.object({
  network: z.string(),
  username: z.string().optional(),
  url: z.string().url(),
});
export type ProfileLink = z.infer<typeof ProfileLink>;

export const Basics = z.object({
  name: z.string(),
  label: z.string().optional(),
  email: z.string().email().optional(),
  phone: z.string().optional(),
  url: z.string().url().optional(),
  summary: z.string().optional(),
  location: Location.optional(),
  profiles: z.array(ProfileLink).default([]),
});
export type Basics = z.infer<typeof Basics>;

export const Work = z.object({
  name: z.string(),
  position: z.string(),
  url: z.string().url().optional(),
  location: z.string().optional(),
  startDate: z.string().optional(),
  endDate: z.string().optional(),
  summary: z.string().optional(),
  highlights: z.array(z.string()).default([]),
  tech: z.array(z.string()).default([]),
});
export type Work = z.infer<typeof Work>;

export const Education = z.object({
  institution: z.string(),
  area: z.string().optional(),
  studyType: z.string().optional(),
  startDate: z.string().optional(),
  endDate: z.string().optional(),
  score: z.string().optional(),
  courses: z.array(z.string()).default([]),
});
export type Education = z.infer<typeof Education>;

export const Skill = z.object({
  name: z.string(),
  level: z.string().optional(),
  keywords: z.array(z.string()).default([]),
});
export type Skill = z.infer<typeof Skill>;

export const Project = z.object({
  name: z.string(),
  description: z.string().optional(),
  url: z.string().url().optional(),
  highlights: z.array(z.string()).default([]),
  keywords: z.array(z.string()).default([]),
});
export type Project = z.infer<typeof Project>;

export const Certificate = z.object({
  name: z.string(),
  date: z.string().optional(),
  issuer: z.string().optional(),
  url: z.string().url().optional(),
});
export type Certificate = z.infer<typeof Certificate>;

export const Resume = z.object({
  basics: Basics,
  work: z.array(Work).default([]),
  education: z.array(Education).default([]),
  skills: z.array(Skill).default([]),
  projects: z.array(Project).default([]),
  certificates: z.array(Certificate).default([]),
  meta: z
    .object({ lastUpdated: z.string().optional(), version: z.string().optional() })
    .optional(),
});
export type Resume = z.infer<typeof Resume>;

// Public-facing alias for the résumé document.
export const Profile = Resume;
export type Profile = Resume;
